src/OmniSharp.MSBuild/ProjectLoadListener.cs

```diff
@@ -0,0 +1,106 @@
+        internal const string TargetFrameworksPropertyName = "TargetFrameworks";
+        private const string MSBuildProjectFullPathPropertyName = "MSBuildProjectFullPath";
+        private readonly ILogger _logger;
+        private readonly IEventEmitter _eventEmitter;
+        private static readonly VsTfmAndFileExtHashingAlgorithm _tfmAndFileHashingAlgorithm = new VsTfmAndFileExtHashingAlgorithm();
+        private static readonly VsReferenceHashingAlgorithm _referenceHashingAlgorithm = new VsReferenceHashingAlgorithm();
+
+        [ImportingConstructor]
+        public ProjectLoadListener(ILoggerFactory loggerFactory, IEventEmitter eventEmitter)
+        {
+            _logger = loggerFactory.CreateLogger<ProjectLoadListener>();
+            _eventEmitter = eventEmitter;
+        }
+
+        public void ProjectLoaded(ProjectLoadedEventArgs args)
+        {
+            try
+            {
+                var projectGuid = GetProjectId(args);
+                var hashedTargetFrameworks = GetHashedTargetFrameworks(args.ProjectInstance);
+
+                if (args.References == null)
+                {
+                    return;
+                }
+
+                var hashedReferences = GetHashedReferences(args);
+                var hashedFileExtensions = GetUniqueHashedFileExtensions(args);
+
+                _eventEmitter.ProjectInformation(projectGuid, hashedTargetFrameworks, hashedReferences, hashedFileExtensions);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError("Unexpected exception got thrown from project load listener: " + ex);
+            }
+        }
+
+        private static IEnumerable<HashedString> GetUniqueHashedFileExtensions(ProjectLoadedEventArgs args)
+        {
+            IEnumerable<string> sourceFileExtensions = args.SourceFiles.Select(file => Path.GetExtension(file)).Distinct();
+            return sourceFileExtensions.Select(ext => _tfmAndFileHashingAlgorithm.HashInput(ext));
+        }
+
+        private static HashedString GetProjectId(ProjectLoadedEventArgs args)
+        {
+            if (args.HasProjectIdGeneratedFromSolution)
+            {
+                //If we are getting a raw guid we should not hash it
+                return new HashedString(args.Id.Id.ToString());
+            }
+
+            var projectFilePath = args.ProjectInstance.GetPropertyValue(MSBuildProjectFullPathPropertyName);
+            var content = File.ReadAllText(projectFilePath);
+            //create a hash from the filename and the content
+            return _referenceHashingAlgorithm.HashInput($"Filename: {Path.GetFileName(projectFilePath)}\n{content}");
+        }
+
+        private static IEnumerable<HashedString> GetHashedReferences(ProjectLoadedEventArgs args)
+        {
+            var referenceNames = args.References.Select(reference => Path.GetFileNameWithoutExtension(reference).ToLower());
```

Comment:
Should we be removing the extension?

Reply:
Should we be tolower?

Reply:
Yeah as per what Andrew said we need to do "system.core"